### PR TITLE
1.2.0 pre-release

### PR DESCRIPTION
**Pull request by BusyBeaver-42**

### CODE REVIEW

File: src/simulated_bifurcation/polynomial/ising_polynomial_interface.py [hunk 1]
```diff
@@ -1,69 +1,130 @@
...
-    
+                value = torch.tensor(value, dtype=self.dtype, device=self.device)
+            except Exception as err:
+                raise TypeError(f"Input value cannot be cast to Tensor.") from err
+        if (self.__accepted_values is not None) and torch.any(
+            torch.isin(value, self.__accepted_values, invert=True)
+        ):
+            raise ValueError(
+                f"Input values must all belong to {self.__accepted_values}."
+            )
```
Comment (bqth29) [inline src/simulated_bifurcation/polynomial/ising_polynomial_interface.py:107]: maybe `self.__accepted_values.tolist()` from better readability

File: src/simulated_bifurcation/__init__.py [hunk 2]
```diff
@@ -1,7 +1,128 @@
...
-from .optimizer import SimulatedBifurcationOptimizer, get_env, reset_env, set_env
-from .polynomial import SpinPolynomial, BinaryPolynomial, IntegerPolynomial
-from .simulated_bifurcation import minimize, maximize
+"""
+Simulated Bifurcation
+=====================
 
+Provides
+  1. GPU compatible implementation of the simulated bifurcation (SB)
+  algorithm, a physics inspired combinatorial approximation algorithm.
```
Comment (bqth29) [inline src/simulated_bifurcation/__init__.py:7]: quantum physics?

File: src/simulated_bifurcation/__init__.py [hunk 3]
```diff
@@ -1,7 +1,128 @@
...
+  discrete SB : Uses the sign of the position of the particles for the
+    position-based update of the momentums ; usually slower but more
+    accurate.
+On top of these two variants, an additional thermal fluctuation term
+can be added in order to help escape local optima [4]. Use this
+additional term by setting `heated=True`.
+
+The hyperparameters of the SB algorithm which correspond to physical
+constants have been fine-tuned  (Goto et al.) to give the best results most
+of the time. Nevertheless, the relevance of specific  hyperparameters may
```
Comment (bqth29) [inline src/simulated_bifurcation/__init__.py:61]: there are two whitespaces before hyperparameters

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 4]
```diff
@@ -1,57 +1,837 @@
...
+    `ΣΣ Q(i,j)x(i)x(j) + Σ l(i)x(i) + c`
+    or `x.T Q x + l.T x + c` in matrix notation,
+    where `Q` is a square matrix, `l` a vector a `c` a constant.
+
+    Parameters
+    ----------
+    matrix : (M, M) Tensor | ndarray
+        Matrix corresponding to the quadratic terms of the polynomial
+        (quadratic form). It should be a square matrix, but it needs not
+        be symmetric.
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:70]: I would say "but not necessarily symmetric"

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 5]
```diff
@@ -1,57 +1,837 @@
...
+            Optimize the polynomial over vectors whose entries are n-bits
+            non-negative integers, that is integers between 0 and 2^n - 1
+            inclusive.
+
+    dtype : torch.dtype, default=torch.float32
+        Data-type used for running the computations in the SB algorithm.
+    device : str | torch.device, default="cpu"
+        Device on which the SB algorithm is run. If available, use "cuda"
+        to run the SB algorithm on GPU (much faster, especially for high
+        dimensional instances or  when running the algorithm with many
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:95]: two whitespaces before when

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 6]
```diff
@@ -1,57 +1,837 @@
...
+    sampling_period : int, default=50
+        Number of iterations between two consecutive spin samplings by the
+        window.
+    convergence_threshold : int, default=50
+        Number of consecutive identical spin samplings considered as a
+        proof of convergence by the window.
+    Hyperparameters corresponding to physical constants :
+        These parameters have been fine-tuned (Goto et al.) to give the
+        best results most of the time. Nevertheless, the relevance of
+        specific  hyperparameters may vary depending on the properties of
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:144]: two whitespaces before hyperparameters

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 7]
```diff
@@ -1,57 +1,837 @@
...
+    `ΣΣ Q(i,j)x(i)x(j) + Σ l(i)x(i) + c`
+    or `x.T Q x + l.T x + c` in matrix notation,
+    where `Q` is a square matrix, `l` a vector a `c` a constant.
+
+    Parameters
+    ----------
+    matrix : (M, M) Tensor | ndarray
+        Matrix corresponding to the quadratic terms of the polynomial
+        (quadratic form). It should be a square matrix, but it needs not
+        be symmetric.
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:300]: I would say "but not necessarily symmetric"

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 8]
```diff
@@ -1,57 +1,837 @@
...
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:521]: I would say "but not necessarily symmetric"

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 9]
```diff
@@ -1,57 +1,837 @@
...
+            non-negative integers, that is integers between 0 and 2^n - 1
+            inclusive.
+
+    dtype : torch.dtype, default=torch.float32
+        Data-type used for running the computations in the SB algorithm.
+    device : str | torch.device, default="cpu"
+        Device on which the SB algorithm is run. If available, use "cuda"
+        to run the SB algorithm on GPU (much faster, especially for high
+        dimensional instances or  when running the algorithm with many
+        agents).
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:326]: two whitespaces before when

File: src/simulated_bifurcation/simulated_bifurcation.py [hunk 10]
```diff
@@ -1,57 +1,837 @@
...
+            Optimize the polynomial over vectors whose entries are n-bits
+            non-negative integers, that is integers between 0 and 2^n - 1
+            inclusive.
+
+    dtype : torch.dtype, default=torch.float32
+        Data-type used for running the computations in the SB algorithm.
+    device : str | torch.device, default="cpu"
+        Device on which the SB algorithm is run. If available, use "cuda"
+        to run the SB algorithm on GPU (much faster, especially for high
+        dimensional instances or  when running the algorithm with many
```
Comment (bqth29) [inline src/simulated_bifurcation/simulated_bifurcation.py:546]: two whitespaces before when

**Review by bqth29 — COMMENTED** [review]
Amazing documentation!

File: src/simulated_bifurcation/models/number_partitioning.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-from typing import Dict, List, Union
+from typing import Dict, List, Tuple, Union
```
Comment (BusyBeaver-42) [inline src/simulated_bifurcation/models/number_partitioning.py:1]: unused import Tuple

**Review by bqth29 — APPROVED** [review]

**Pull request merged**